internal/js_parser/js_parser.go

```diff
@@ -1733,6 +1741,297 @@ func (p *parser) importJSXSymbol(loc logger.Loc, jsx JSXImport) js_ast.Expr {
 	})
 }
 
+func isValidReactComponentName(candidate string) bool {
+	return len(candidate) > 0 && candidate[0] >= 'A' && candidate[0] <= 'Z'
+}
+
+func isReactCustomHook(candidate string) bool {
+	return len(candidate) >= 4 &&
+		candidate[0] == 'u' &&
+		candidate[1] == 's' &&
+		candidate[2] == 'e' &&
+		candidate[3] >= 'A' &&
+		candidate[3] <= 'Z'
+}
+
+var reactBuiltInHooks = map[string]bool{
+	"useState":            true,
+	"useReducer":          true,
+	"useEffect":           true,
+	"useLayoutEffect":     true,
+	"useMemo":             true,
+	"useCallback":         true,
+	"useRef":              true,
+	"useContext":          true,
+	"useImperativeHandle": true,
+	"useDebugValue":       true,
+}
+
+type reactRefreshHandle struct {
+	handle ast.Ref
+	id     string
+}
+
+func (p *parser) reactRefreshRegisterHandle(target ast.Ref, id string) js_ast.Stmt {
+	var handleName = "_c"
```

Comment:
I think it would be good to handle scope collisions (like the SWC transform):

```js
const _c = 'collision';
const _c1 = 'collision1';
export function Component() {
  return <div></div>;
}

// becomes:

import { jsxDEV as o } from "react/jsx-dev-runtime";
const _c = 'collision';
const _c1 = 'collision1';
export function Component() {
    return /*#__PURE__*/ o("div", {}, void 0, false, {
        fileName: "input.jsx",
        lineNumber: 4,
        columnNumber: 10
    }, this);
}
_c2 = Component;
var _c2;
$RefreshReg$(_c2, "Component");
```

In the case of SWC, the integer suffix is incremented until a unique identifier is found in scope.

Reply:
@ArnaudBarre any update?

Reply:
Nothing plan for the moment until I get more positive feedback on this from either maintainers or the community. 16 reactions for something that would allow instant dev server for SPA in dev (because not other tools can do bundling with fast refresh without plugin overhead) feels not worth it publishing this as a fork